src/vos/vos_space.c

```diff
@@ -303,19 +303,42 @@ vos_space_hold(struct vos_pool *pool, uint64_t flags, daos_key_t *dkey,
 	if (scm_left < space_est[DAOS_MEDIA_SCM])
 		goto error;
 
-	/* If NVMe isn't configured or this update doesn't use NVMe space */
-	if (pool->vp_vea_info == NULL || space_est[DAOS_MEDIA_NVME] == 0)
-		goto success;
+	/* If NVMe is configured and this update uses NVMe space */
+	if (pool->vp_vea_info != NULL && space_est[DAOS_MEDIA_NVME] != 0) {
+		nvme_left = NVME_FREE(&vps);
 
-	nvme_left = NVME_FREE(&vps);
-	if (nvme_left < NVME_SYS(&vps))
-		goto error;
+		if (nvme_left < NVME_SYS(&vps))
+			goto error;
 
-	nvme_left -= NVME_SYS(&vps);
-	/* 'NVMe held' has already been excluded from 'NVMe free' */
+		nvme_left -= NVME_SYS(&vps);
+		/* 'NVMe held' has already been excluded from 'NVMe free' */
 
-	if (nvme_left < space_est[DAOS_MEDIA_NVME])
-		goto error;
+		if (nvme_left < space_est[DAOS_MEDIA_NVME])
+			goto error;
+	}
+
+	/* Check space reserve for rebuild */
```

Comment:
This is to make sure the non-rebuild IO will not use the space to reserve for rebuild? I assume other background ULT will need reserve the space as well? 
So maybe use general name here? (I mean VOS_OF_REBUILD and vp_space_rb).

Reply:
This check is for rebuild only and this vos_space_hold() is called by update request processing only.

Reply:
I mean it can be 

if (!(flags & VOS_USE_RESERVE_SPC) {

}. 
Since these seems very general checking, nothing related with the rebuild except the flag?

And also the comments is a bit confused, since this is actually for checking no-rebuild case? These are not logical fault.

Reply:
This isn't the general checking for "sys reserve", this is the special checking for "rebuild reserve".

"sys reserve" is to reserve space for truncate/destroy/aggregation etc. It's for ensuring the space reclaim can still work even when free space is tight, "sys reserve" is mandatory and not configurable.

"rebuild reserve" is to reserve space for potential rebuild purpose, it's configurable. A user who doesn't use redundant objects should obviously configure the "rebuild reserve" as zero for better space utilization.